### PR TITLE
[Topi] Add batch_gather op

### PR DESCRIPTION
**Pull request by hlu1**
Implement batch_gather op as in caffe2: https://github.com/pytorch/pytorch/blob/master/caffe2/operators/batch_gather_ops.cc#L24-L43

@ajtulloch, @antinucleon please review.

### CODE REVIEW

**Comment by ajtulloch** [comment]
It's also identical to [`tf.batch_gather`](https://www.tensorflow.org/api_docs/python/tf/batch_gather) right? Looks good to me.

**Comment by hlu1** [comment]
Talked to @antinucleon offline. We realized that the c2 `batch_gather` is the same as the tvm `take` op with `axis=1 and mode="clip"`. However, the tvm version is about 30% slower because it clipped the indices (see https://github.com/dmlc/tvm/blob/master/topi/include/topi/transform.h#L701). I'm thinking about adding another "fast" mode to the tvm take op, similar to the numpy "raise" mode (https://docs.scipy.org/doc/numpy/reference/generated/numpy.take.html).

**Comment by hlu1** [comment]
RFC https://github.com/dmlc/tvm/issues/3303
